### PR TITLE
[SPARK-6492][CORE] SparkContext.stop() can deadlock when DAGSchedulerEventProcessLoop dies

### PR DESCRIPTION
**Pull request by ilganeli**
I've added a timeout and retry loop around the SparkContext shutdown code that should fix this deadlock. If a SparkContext shutdown is in progress when another thread comes knocking, it will wait for 10 seconds for the lock, then fall through where the outer loop will re-submit the request.

### CODE REVIEW

**Comment by JoshRosen** [comment]
(Cross-posted from JIRA):

Timeouts are one way to fix this, but I wonder if we could also try to remove the circular wait condition by modifying EventLoop so that `stopped` is set before we call `onError`. This would prevent calls to EventLoop.stop() from blocking while the event loop is in the process of shutting down, which should prevent this race.

/cc @zsxwing, who wrote `EventLoop`.

**Comment by shaneknapp** [comment]
whoops, my bad.  didn't mean to kill this job.

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by ilganeli** [comment]
@JoshRosen  - there's a number of places throughout the code ```(SparkContext.stop(), ContextCleaner.stop(), SparkEnv.stop()``` that I've found so far where the stopped boolean is set before the stop sequence is complete. As I mention in the intro for this PR, this can introduce a host of problems with incomplete shutdowns. I'd like to clean this up throughout the code base and would prefer to avoid adding more instances of this practice.

**Comment by ilganeli** [comment]
retest this please

**Comment by JoshRosen** [comment]
Hmm, good points.

In the case of the race reported in the original JIRA, I guess the main thread's `SparkContext.stop()` call wins, so the EventLoop thread would lose and throw the timeout exception, unblocking the main thread and allowing it to finish cleanup.  If the locks were acquired in the other order, then I guess that the EventLoop would wind up trying to `join()` on its own thread.  Based on http://stackoverflow.com/questions/5999595/thread-join-on-itself, it sounds like this is not safe.  Do you think that we should guard against this by checking the current thread id in `EventLoop.stop()` before calling join()?

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -1394,28 +1396,37 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
...
-        SparkContext.clearActiveContext()
-      } else {
-        logInfo("SparkContext already stopped")
+    // The shutdown logic can create a deadlock scenario when an external thread triggers an
+    // error in the handling of which it shuts down the SparkContext. Meanwhile, this thread
+    // will wait for the external thread to join during the "dagScheduler.stop()" while that
+    // thread waits on the SPARK_CONTEXT_CONSTRUCTOR_LOCK. We avoid this by surrounding this
+    // logic in a try catch.
+    while (!stopped) {
+      if (SparkContext.shutdownLock.tryLock(10, TimeUnit.SECONDS)) {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: I think we need a `finally` to ensure that the lock acquired here is eventually released.

Comment (ilganeli) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: I'll add that in.

**Comment by zsxwing** [comment]
From the comments, only  `activeContext`, `contextBeingConstructed` are protected by `SPARK_CONTEXT_CONSTRUCTOR_LOCK`. So we can move most of codes out of the lock. E.g.,

```Scala
  def stop() {
    SparkContext.SPARK_CONTEXT_CONSTRUCTOR_LOCK.synchronized {
      if (!stopped) {
        stopped = true
      } else {
        logInfo("SparkContext already stopped")
        return
      }
    }
    postApplicationEnd()
    ui.foreach(_.stop())
    env.metricsSystem.report()
    metadataCleaner.cancel()
    cleaner.foreach(_.stop())
    dagScheduler.stop()
    dagScheduler = null
    listenerBus.stop()
    eventLogger.foreach(_.stop())
    env.rpcEnv.stop(heartbeatReceiver)
    progressBar.foreach(_.stop())
    taskScheduler = null
    // TODO: Cache.stop()?
    env.stop()
    SparkEnv.set(null)
    logInfo("Successfully stopped SparkContext")
    SparkContext.clearActiveContext()
  }
```

**Comment by zsxwing** [comment]
@JoshRosen I assume these `stop`s should be thread-safe. Do you know which is not?

**Comment by zsxwing** [comment]
> Timeouts are one way to fix this, but I wonder if we could also try to remove the circular wait condition by modifying EventLoop so that stopped is set before we call onError. This would prevent calls to EventLoop.stop() from blocking while the event loop is in the process of shutting down, which should prevent this race.

Do you mean calling `EventLoop.stop` in dagScheduler.onError? I just found that calling `EventLoop.stop` in `EventLoop.onReceive` and `EventLoop.onError` won't call onStop. I will fix it later.

**Comment by zsxwing** [comment]
Fix the `EventLoop.stop` issue in #5280

**Comment by JoshRosen** [comment]
> Do you mean calling EventLoop.stop in dagScheduler.onError?

I was referring to the race that I reported in [SPARK-6492](https://issues.apache.org/jira/browse/SPARK-6492) (referred to by this PR), where the EventLoop.onError() calls SparkContext.stop(), which blocks on the SPARK_CONTEXT_CONSTRUCTOR_LOCK lock, while another thread has simultaneously called SparkContext.stop(), acquired the SPARK_CONTEXT_CONSTRUCTOR_LOCK, and called `DAGScheduler.stop()`, which calls `EventLoop.stop()`.  My comment was referring to this `EventLoop.stop()` call blocking indefinitely while waiting to `join()` on the event processing thread, which is blocked on acquiring the `SPARK_CONTEXT_CONSTRUCTOR_LOCK`.

Based on the comments upthread from @ilganeli, I don't think that we should adopt my earlier suggestion of having `EventLoop.stop()` being a no-op as long as some other thread is in the process of stopping the EventLoop, since this could result in a scenario where in-progress cleanup is still happening after a user's call to SparkContext.stop() has returned, which could lead to cleanup being skipped were the JVM to exit at that point.

I am worried about the case where calling `EventLoop.stop()` from the event loop thread itself (which can happen transitively here) leads to a one-threaded deadlock, but I guess your other patch addresses this?

**Comment by zsxwing** [comment]
> I am worried about the case where calling EventLoop.stop() from the event loop thread itself (which can happen transitively here) leads to a one-threaded deadlock, but I guess your other patch addresses this?

Yes.  Calling EventLoop.stop() from the event loop thread itself won't lead a one-threaded deadlock.

@JoshRosen Your thoughts about moving these `stop`s out of the lock?

**Comment by JoshRosen** [comment]
> From the comments and codes, only activeContext, contextBeingConstructed and stopped are protected by SPARK_CONTEXT_CONSTRUCTOR_LOCK.

@zsxwing, I took a closer look and I'm not even sure that we need the synchronization for guarding `stopped`, since it's volatile: 

```
  @volatile private var stopped: Boolean = false
```

I suppose that the `SPARK_CONTEXT_CONSTRUCTOR_LOCK` was preventing multiple threads from being in `stop()` at the same time, but I don't know that we need to enforce that requirement; it's not enforced in pre-1.2 versions of the code (such as https://github.com/apache/spark/blob/39761f515d65afff377873ee4701b9313c317a60/core/src/main/scala/org/apache/spark/SparkContext.scala#L1024).  Maybe we can resolve this deadlock by removing this `synchronized` call?

**Comment by JoshRosen** [comment]
In fact, it looks like the `synchronized` call in `stop()` was added in a very early iteration on that patch but became unnecessary after the code was refactored during code review.

**Comment by zsxwing** [comment]
> Maybe we can resolve this deadlock by removing this synchronized call?

Removing `synchronized` is not enough. `stopped` also need to be AtomicBoolean to make the reading and writing `stopped` atomic.

**Comment by JoshRosen** [comment]
> `stopped` also need to be AtomicBoolean to make the reading and writing stopped atomic.

Ah, right, since we're doing a compare-and-swap here (I got mixed up thinking of a different patch; sorry).

**Comment by ilganeli** [comment]
Ok so it looks like we have a simple fix for the time being, in the long-term, so we aren't assigning to the ```stopped``` before the entire process is complete, I'd prefer a slightly different solution. I would add a second AtomicBoolean ```stopping``` and make the if check to be ```if(!stopping && !stopped)```. Stopping would be set true in the beginning of the process and stopped would be set true at the end. This would guard against incomplete stops.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -1791,6 +1804,11 @@ object SparkContext extends Logging {
   private val SPARK_CONTEXT_CONSTRUCTOR_LOCK = new Object()
 
   /**
+   * Lock to guard against deadlock when shutting down SparkContext.
+   */
+  private val shutdownLock = new ReentrantLock()
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Looks like this is now unused.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -95,10 +97,11 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   val startTime = System.currentTimeMillis()
 
-  @volatile private var stopped: Boolean = false
+  @volatile private var stopped: AtomicBoolean = new AtomicBoolean(false)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: You can remove `@volatile` and use `val` for AtomicBoolean

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 4]
```diff
@@ -1394,30 +1397,40 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
...
-        taskScheduler = null
-        // TODO: Cache.stop()?
-        env.stop()
-        SparkEnv.set(null)
-        logInfo("Successfully stopped SparkContext")
-        SparkContext.clearActiveContext()
+    // Use the stopping variable to ensure no contention for the stop scenario.
+    // Still track the stopped variable for use elsewhere in the code.
+    if (!stopped.get()) {
+      if(!stopping.get()) {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Here should be `stopping.compareAndSet(false, true)`

**Comment by zsxwing** [comment]
Do we really need both `stopping` and `stopped`? From the usage of `assertNotStopped()`, I think if `sc.stop()` is called, calling other methods of `sc` should be rejected.

**Comment by JoshRosen** [comment]
Woah, very weird MiMa error:

```
[error]  * synthetic method org$apache$spark$SparkContext$$SPARK_CONTEXT_CONSTRUCTOR_LOCK()java.lang.Object in object org.apache.spark.SparkContext does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.SparkContext.org$apache$spark$SparkContext$$SPARK_CONTEXT_CONSTRUCTOR_LOCK")
```

**Comment by ilganeli** [comment]
@zsxwing My reasoning for having both is not to allow other methods of sc to be called but to address the scenario where the context doesn't shut down completely during the shutdown step. ```stopping``` provides the synchronization and ```stopped``` is only set once all is complete.

**Comment by zsxwing** [comment]
LGTM

**Comment by ilganeli** [comment]
Any notion of what's going on with this MiMa failure?

**Comment by zsxwing** [comment]
@ilganeli Could you add 
```
ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.SparkContext.org$apache$spark$SparkContext$$SPARK_CONTEXT_CONSTRUCTOR_LOCK")
```
to `MimaExcludes.scala`

**Comment by ilganeli** [comment]
Certainly - just trying to understand what it's complaining about.

**Comment by zsxwing** [comment]
Before your change, the java interface of SparkContext$ contains:
```
public final class org.apache.spark.SparkContext$ extends java.lang.Object implements org.apache.spark.Logging{
    public java.lang.Object org$apache$spark$SparkContext$$SPARK_CONTEXT_CONSTRUCTOR_LOCK();
}
```
But after your change, SPARK_CONTEXT_CONSTRUCTOR_LOCK isn't used out of `SparkContext$`, so the compiler does not generate 
```
public java.lang.Object org$apache$spark$SparkContext$$SPARK_CONTEXT_CONSTRUCTOR_LOCK();
```

I think it's OK since this is not a Scala public API.

**Comment by srowen** [comment]
LGTM. I'll wait a bit for more comments.

**Comment by srowen** [comment]
Any other comments? I'd like to merge today.

**Comment by zsxwing** [comment]
:+1:

**Comment by srowen** [comment]
Ack, sorry @ilganeli but looks like this needs a little rebase. It's ready to go then.

**Comment by ilganeli** [comment]
Fixed. Thanks.

**Comment by ilganeli** [comment]
retest this please

**Pull request closed without merging**